### ISSUE TITLE
helm dependency update fails to get the correct version when using 0.1.0+<commit#>.<githash> versioning

### ISSUE THREAD

**Issue by noam-toga** [helm/helm#7563]
In our workflow, a developer tags a version of Helm chart like this
0.1.0+sequentional-num.git-hash
once pushed we can have in our repo 2 versions of helm charts like this

chartX	0.5.0+1.69b12cf
chartX	0.5.0+2.bbfb1e3   

Seems like when i change my requirements.yaml between the 2 versions above amd run 'dep build' and 'dep update'
1) requirements.lock doesnt change or re-generated with the version in the requirements.lock (it keeps one of those versions forever)
2) dep update brings the same tgz as always and never fetches the right version from the requirements.yaml
                         
Seems like the logic is only to take the 0.5.0 and ignore the suffix. 
This is very crucial for us. as many developers works on 0.5.0 and before changing the version we need to stick their git hashes in their version.
Any idea how can i solve this?

-------------- Workspace -------------------
Output of `helm version`:
version.BuildInfo{Version:"v3.0.2", GitCommit:"19e47ee3283ae98139d98460de796c1be1e3975f", GitTreeState:"clean", GoVersion:"go1.13.5"}

Output of `kubectl version`:
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T19:18:23Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-12-04T07:23:47Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
kind

**Comment by mattfarina** [comment]
I think I know what's going on here.

The `+` signifies build metadata. The idea is that version `0.5.0+1.69b12cf` and `0.5.0+2.bbfb1e3` are seen as equivalent and two different builds of the same version `0.5.0`. This is how semver works. Specifically, the [specification says](https://semver.org/#spec-item-10) (emphasis mine):

> Build metadata MAY be denoted by appending a plus sign and a series of dot separated identifiers immediately following the patch or pre-release version. Identifiers MUST comprise only ASCII alphanumerics and hyphen [0-9A-Za-z-]. Identifiers MUST NOT be empty. Build metadata MUST be ignored when determining version precedence. **Thus two versions that differ only in the build metadata, have the same precedence.** Examples: 1.0.0-alpha+001, 1.0.0+20130313144700, 1.0.0-beta+exp.sha.5114f85.

There are a couple options you can use for Helm to fix this.

1. You can use pre-releases instead. Those are to be used in ordering. [You can read about them in the spec](https://semver.org/#spec-item-9). In `requirements.yaml` file make sure you are looking for pre-releases. The spec notes that they may not have the same API as the final release so we filter them out when only looking for stable releases. If you use releases like `0.5.0-1.69b12cf` and `0.5.0-2.bbfb1e3` these will be seen as coming before `0.5.0` and `2` as newer than `1`. In the `requirements.yaml` file do something like `>0.5-0`. Ordering uses [ascii](http://www.asciitable.com/) so A-Z comes before a-z (not equal fold).
2. In the `requirements.yaml` file you can specify a specific version (e.g., `0.5.0+2.bbfb1e3`). Helm will fetch that if available. You will need to update the `requirements.yaml` file for each update rather than relying on Helm to do it for the lock file.

This all has to do with how semantic versioning works.

**Comment by noam-toga** [comment]
@mattfarina Thank you for your detailed answer. i probably didnt explain my issue right. let me try again.
tl;dr - i wish your 2nd suggestion would have worked. this could've solved all of our problems :))

So i have 2 versions of my helmchart called dnsutils in my helm repo.
output of `helm search repo -l`:

`NAME                      	CHART VERSION  	APP VERSION	DESCRIPTION                                       
chartmuseum/dnsutils      	0.1.0          	        1.3        	A Helm chart for Kubernetes                       
chartmuseum/dnsutils      	0.1.0+1.abcdeff	1.3        	A Helm chart for Kubernetes`

my simple `requirements.yaml`:

`dependencies:
    - name: dnsutils
      version: "0.1.0+1.abcdeff"
      repository: "@chartmuseum"
      condition: dnsutils.enabled
      tags:
        - debug
`

now when i try to get the dependencies for that yaml with helm dep cmd - 
`helm dependency update charts/workspaces/debug`

i get the wrong version! i get `dnsutils-0.1.0.tgz` instead of  `dnsutils-0.1.0+1.abcdeff.tgz`

output of the helm dep update directory - 
`ls charts/workspaces/debug/charts/` - 

`dnsutils-0.1.0.tgz`

it's very easy to reproduce.

Let me know if i was understood this time
And thanks again for the help
Noam

**Comment by technosophos** [comment]
@mattfarina can you provide an update on this? I can't tell if this is a bug, or if there is some nuance to SemVer that I am not understanding here. But it seems to me that this is a bug.

**Comment by belolap** [comment]
In prev versions it was a bug, that was fixed: #4838 (as referenced upper). Do something changed in SemVer interpretation?

**Comment by bacongobbler** [comment]
https://github.com/helm/helm/pull/6821 ported the fix over from helm 2. It was patched in helm 3.1.0. Which version are you testing with?

**Comment by belolap** [comment]
In my CI/CD scripts uses v3.2.1. I'll double check all now.

**Comment by belolap** [comment]
Yes, version 3.2.1:

```
> helm version
version.BuildInfo{Version:"v3.2.1", GitCommit:"fe51cd1e31e6a202cba7dead9552a6d418ded79a", GitTreeState:"clean", GoVersion:"go1.13.10"}

> grep -A 2 "name: house" Chart.yaml 
  - name: house
    version: 1.0.0-dev5+11.7ad65ae4
    repository: "@bigur"

> helm search repo house --devel -l
NAME            CHART VERSION           APP VERSION             DESCRIPTION                                
bigur/house     1.0.0-dev5+11.7ad65ae4  1.0.0-dev5+11.7ad65ae4  Микросервис с информацией о пользователях
bigur/house     1.0.0-dev5              1.0.0-dev5              Микросервис с информацией о пользователях

> helm pull bigur/house --version 1.0.0-dev5+11.7ad65ae4 && ls *.tgz
house-1.0.0-dev5+11.7ad65ae4.tgz

> helm dependency update && ls -lFh charts
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "bigur" chart repository
Update Complete. ⎈Happy Helming!⎈
Saving 3 charts
Downloading house from repo http://charts.default.svc.k.bigur.lan:8080/
Deleting outdated charts
итого 12K
-rw-r--r-- 1 gik gik 2,0K июн 23 16:41 house-1.0.0-dev5.tgz
```

**Comment by belolap** [comment]
When we build development version we was use this format of version:

```
<release>-<suffix>+<commits>.<sha>
```
Where release is the future release number of product, suffix one of `-dev`, `-alpha`, `-beta` and so on, commits - number of commits in branch after last tag and sha is last commit't sha.

I have discussion with frieds and we deside that use number of commits in version's metadata is incorrect. Content of packages when versions before `+` match must be same. So if we have version `1.0.0-dev+10` and `1.0.0-dev+11` and content of packages different it is incorrect interpretation of semver.

Now we move number of commits before `+` sign:

```
<release>-<suffix>.<branch_sha>.<commits>+<sha>
```

And we calculate sha of current branch name and take first 6 chars of it and put it as `branch_sha` if branch name is not master. It is added for differ in which branch package was created. In this scheme `+sha` is unneeded for my case, but we keep it for some control (for example if was used `git rebase`).

So now I have no problems with helm dependency update.

**Comment by bacongobbler** [comment]
> So now I have no problems with helm dependency update.

I’m confused. Are you suggesting that this is fixed, or that you found a workaround? I’m not sure if this means the ticket can be closed or something else.

**Comment by belolap** [comment]
I am not author of the ticket, so can't say can be ticket closed or not. 

My oppinion:

1. I can't found usecase when helm dependency update must download package with some special metadata in version because content of packages with different metadata and same version must be same. In my usecase that I say before was mistake to understand of semver standard. So in one hand ticket can be closed.

2. In another hand issue must be resolved because it is user friendly, user may want to have possibility to point exact version of package. And other subcommands of helm can point for specific version by `--version flag`.

So if it is not much work better to resolve it.

**Comment by imanushin** [comment]
To continue the discussion.

I found that `chartrepo.go` has the following code:
```go
cv, err := repoIndex.Get(chartName, chartVersion)
```

And `index.go` does exact match if it can. Therefore it seems like if version `X` was requested and `X` is available in index then `X` should be selected, not `X+abc`.

I created PR #11502, however it seems like helm selects exact version it is was requested though command line.

Question: if repository has versions `1.0.0` and `1.0.0+abc` for chart A, how it is possible to have it as a dependency in chart B and request `1.0.0` version (not `1.0.0+abc` which we have for now)?
